File: Loenn/entities/coin.lua
local coin = {}

coin.fieldInformation = {
   variant = {
      options = {
         "yellow", "blue",
      },
      editable = false,
   },
}

coin.name = "Casju0TrailMix/Coin"
coin.depth = 2000
coin.placements = { {
   name = "coin_yellow",
   data = {
      variant = "yellow",
   },
}, {
   name = "coin_blue",
   data = {
      variant = "blue",
   },
} }

function coin.texture(room, entity)
   return "objects/Casju0TrailMix/coin/" .. entity.variant .. "00"
end

return coin

File: Source/Entities/Coin.cs
namespace Celeste.Mod.Casju0TrailMix.Entities;

[Tracked(false)]
[CustomEntity("Casju0TrailMix/Coin")]
public class Coin : Entity
{
    private readonly Hitbox hitbox;
    private readonly Sprite sprite;

    public Coin(EntityData data, Vector2 offset)
        : base(data.Position + offset)
    {
        var variant = data.Attr("variant", "yellow");
        Add(sprite = new Sprite(GFX.Game, "objects/Casju0TrailMix/coin/"));
        sprite.AddLoop(
            "idle",
            0.136f,
            GFX.Game.GetAtlasSubtextures($"objects/Casju0TrailMix/coin/{variant}").ToArray()
        );
        sprite.Add(
            "collect",
            0.017f,
            "done",
            GFX.Game.GetAtlasSubtextures($"objects/Casju0TrailMix/coin/sparkle").ToArray()
        );
        sprite.AddLoop(
            "done",
            0f,
            GFX.Game.GetAtlasSubtextures($"objects/Casju0TrailMix/coin/empty").ToArray()
        );
        sprite.SetOrigin(8f, 11f);
        sprite.Play("idle");

        Depth = Depths.Below;
        Collidable = true;
        Visible = true;
        Collider = hitbox = new Hitbox(12f, 16f, -6f, -8f);

        Add(new PlayerCollider(OnCollide, hitbox));
    }

    private void OnCollide(Player player)
    {
        Audio.Play("event:/casju0_TrailMix/smw_coin", Position);
        sprite.Play("collect");
        Collidable = false;
    }
}
